packages/matter.js/src/ble/BtpSessionHandler.ts

```diff
@@ -143,44 +189,135 @@ class BtpSessionHandler {
     async sendMatterMessage(data: ByteArray) {
         logger.debug(`Got Matter message to send via BLE transport: ${data.toHex()}`);
 
-        // TODO implement message assembly
-        // TODO Implement client window handling
-
         if (data.length === 0) { // Just dummy for now that the data parameter is used and not marked unused
             throw new Error("BTP packet must not be empty");
         }
 
-        // Hacky code from here that needs to be completely reworked!
         const header = {
             isHandshakeRequest: false,
             hasManagementOpcode: false,
-            hasAckNumber: this.lastIncomingSequenceNumber !== undefined,
+            hasAckNumber: false,
             isBeginningSegment: true,
-            isEndingSegment: true,
+            isEndingSegment: false,
         };
 
-        // BTP Header is always 3 bytes
-        // 4 when ackNumber is present
-        // 6 when ackNumber and messageLength (aka isBeginningSegment) are present
-        const expectedBtpHeaderLength = 3 + (this.lastIncomingSequenceNumber !== undefined ? 1 : 0) + (header.isBeginningSegment !== undefined ? 2 : 0);
+        // BTP Header is always 2 bytes - flags, sequence number
+        // 3 when ackNumber is present
+        // 5 when ackNumber and messageLength (aka isBeginningSegment) are present
+        let expectedBtpHeaderLength = 2 + (this.serverAckedSequenceNumber !== undefined ? 1 : 0) + (header.isBeginningSegment !== undefined ? 2 : 0);
```

Comment:
Ok lets assume there is alwys an ack nunmber and remove that  here:

```suggestion
        let expectedBtpHeaderLength = 3 + (header.isBeginningSegment !== undefined ? 2 : 0);
```

Reply:
why are we assuming that we will always have the ack number?

Reply:
In this case it is simplification ... because we split the matter message into segments directly and sending later. so there COULD be an ack number (or not). In fact the only difference is that our segments are potentially (if we have no ack number) one byte smaller. But in fact checking at the timepoint of splitting if we "NOW have one ack number to send" is wrong because what if not?

So this is a pragmatic way with the risk of loosing one byte in segment size.

If you want to implement it formally right you need to move the whole packet generation into the "sendQueue" method and remember if you have started a packet already and such and just pushing the matter message into the queue.
Then you do that in that function exactly then when you sent out the packet and the n you know exactly IF you need to add an ack number or not. We culd also adjust it that way

Reply:
okay, so since we are adding the ackNumber later after splitting the data, we will need a buffer of 1 byte for the ackNumber.